packages/flutter/lib/src/widgets/asset_vendor.dart

```diff
@@ -29,12 +30,34 @@ class _ResolvingAssetBundle extends CachingAssetBundle {
   final AssetBundle bundle;
   final _AssetResolver resolver;
 
-  Map<String, String> _keyCache = <String, String>{};
+  Map<String, String> keyCache = <String, String>{};
 
   Future<core.MojoDataPipeConsumer> load(String key) async {
-    if (!_keyCache.containsKey(key))
-      _keyCache[key] = await resolver.resolve(key);
-    return await bundle.load(_keyCache[key]);
+    if (!keyCache.containsKey(key))
+      keyCache[key] = await resolver.resolve(key);
+    return await bundle.load(keyCache[key]);
+  }
+}
+
+// Asset bundle that understands how specific asset keys represent image scale.
+class _ResolutionAwareAssetBundle extends _ResolvingAssetBundle {
+  _ResolutionAwareAssetBundle({
+    AssetBundle bundle,
+    _ResolutionAwareAssetResolver resolver
+  }) : super(
+    bundle: bundle,
+    resolver: resolver
+  );
+
+  _ResolutionAwareAssetResolver get resolver => super.resolver;
+
+  Future<ui.Image> fetchImage(String key) async {
+    core.MojoDataPipeConsumer pipe = await load(key);
+    // At this point the key should be in our key cache, and the image
+    // resource should be in our image cache
+    double scale = resolver.getScale(keyCache[key]);
+    this.imageCache[key].scale = scale;
```

Comment:
Not redundant here, there's a global called "imageCache" that gets precedence if I take away the "this".